tests/test_generate_schema.py

```diff
@@ -456,6 +458,38 @@ def test_all_data_chunks(self):
                     raise e
 
     def verify_data_chunk(self, chunk):
+
```

Comment:
The changes to `generate_schema.py` look great.

But here, Lines 461-525: The problem with using an inner function is that it is so tightly coupled to the outer function that this becomes difficult to maintain and incredibly hard to refactor in the future. I think we can do a lot better by avoiding an inner function. Just extracting the common code into a helper method at the same level, and call the helper method twice. I think you will find that the `git diff` is a lot smaller, and you won't run into wrapping problems, because you avoid indenting huge amount of code.

After you make this edit, you can rebase and squash your 3 commits into a single commit, so that the git diff is cleaner?

```python
    def verify_data_chunk(self, chunk):
      self.verify_data_chunk_as_json_or_dict(chunk=chunk, as_dict=False)
      self.verify_data_chunk_as_json_or_dict(chunk=chunk, as_dict=True)

    def verify_data_chunk_as_json_or_dict(self, *, chunk, as_dict):
        """Verify the given chunk from the testdata.txt file. If `as_dict` is
        True, then if the input_format of the chunk is 'json', then pretend
        that the input data was given as an internal Python dict, and verify
        the 'input_format=dict' code path in SchemaGenerator.
        """
        chunk_count = chunk['chunk_count']
        line_number = chunk['line_number']
        data_flags = chunk['data_flags']
        input_format = 'csv' if ('csv' in data_flags) else 'json'
        keep_nulls = ('keep_nulls' in data_flags)
        infer_mode = ('infer_mode' in data_flags)
        quoted_values_are_strings = ('quoted_values_are_strings' in data_flags)
        sanitize_names = ('sanitize_names' in data_flags)
        ignore_invalid_lines = ('ignore_invalid_lines' in data_flags)
        records = chunk['records']
        expected_errors = chunk['errors']
        expected_error_map = chunk['error_map']
        expected_schema = chunk['schema']
        existing_schema = chunk['existing_schema']

        if as_dict:
          if input_format == 'json':
            print(
                f"Test chunk: {chunk_count}; line_number: {line_number}; "
                f"input_format='dict'"
            )
            input_format = 'dict'
            records = json_reader(records)
          else:
            # Don't bother converting CSV data chunks into Python dict.
            return
        else:
          print(
              f"Test chunk: {chunk_count}; line_number: {line_number}; "
              f"first record: {records[0]}"
          )

        # Generate schema.
        generator = SchemaGenerator(
            input_format=input_format,
            infer_mode=infer_mode,
            keep_nulls=keep_nulls,
            quoted_values_are_strings=quoted_values_are_strings,
            sanitize_names=sanitize_names,
            ignore_invalid_lines=ignore_invalid_lines)
        existing_schema_map = None
        if existing_schema:
            existing_schema_map = bq_schema_to_map(json.loads(existing_schema))
        schema_map, error_logs = generator.deduce_schema(
            records, schema_map=existing_schema_map)
        schema = generator.flatten_schema(schema_map)

        # Check the schema, preserving order
        expected = json.loads(expected_schema, object_pairs_hook=OrderedDict)
        self.assertEqual(expected, schema)

        # Check the error messages
        try:
            self.assertEqual(len(expected_errors), len(error_logs))
        except AssertionError as e:
            print(f"Number of errors mismatched, expected:"
                  f" {len(expected_errors)} got: {len(error_logs)}")
            print(f"Errors: {error_logs}")
            print(f"Expected Errors: {expected_errors}")
            raise e
        self.assert_error_messages(expected_error_map, error_logs)
```

Reply:
Fixed. Renamed `verify_data_chunk_as_json_or_dict ` to `verify_data_chunk_as_csv_json_dict` since it tests csv also.
Do you want me to rebase onto the master branch and open a PR for the master branch, or develop branch?

Reply:
This looks good, thanks for the cleanup. The renaming of the function is ok by me.

I thought that you could squash your 3 commits on your own branch into a single commit, then do a `git push -f` on your own branch, which I think would cause GitHub to pick up the new changes into this PR. But, since this PR is so simple, I think I can just do a squash merge this into develop myself.